crates/orderbook/src/orderbook.rs

```diff
@@ -403,6 +406,66 @@ impl Orderbook {
             .await
             .context("get_user_orders error")
     }
+
+    pub async fn get_order_status(&self, uid: &OrderUid) -> Result<Option<dto::order::Status>> {
+        match self.database.latest_order_event(uid).await? {
+            None => Ok(None),
+            Some(event) => {
+                let fetch_solutions = || async move {
+                    let competition = match self.database.load_latest_competition().await {
+                        Ok(competition) => competition.common,
+                        Err(LoadSolverCompetitionError::NotFound) => {
+                            tracing::warn!("no latest competition exists");
+                            return Ok(Vec::new());
+                        }
+                        Err(err) => {
+                            tracing::error!(?err, "could not load latest competition");
+                            return Ok(Vec::new());
+                        }
+                    };
+                    let solutions = competition
+                        .solutions
+                        .into_iter()
+                        .map(|solution| {
+                            let executed_amounts = solution.orders.iter().find_map(|o| match o {
+                                solver_competition::Order::Legacy { .. } => None,
+                                solver_competition::Order::Colocated {
+                                    id,
+                                    sell_amount,
+                                    buy_amount,
+                                } => (id.0 == uid.0).then_some(dto::order::ExecutedAmounts {
+                                    sell: *sell_amount,
+                                    buy: *buy_amount,
+                                }),
+                            });
+                            dto::order::SolutionInclusion {
+                                solver: solution.solver,
+                                executed_amounts,
+                            }
+                        })
+                        .collect::<Vec<_>>();
+
+                    Ok::<Vec<_>, anyhow::Error>(solutions)
+                };
+                let status = match event.label {
+                    OrderEventLabel::Ready => dto::order::Status::Active,
+                    OrderEventLabel::Created => dto::order::Status::Scheduled,
+                    OrderEventLabel::Considered => {
+                        dto::order::Status::Solved(fetch_solutions().await?)
+                    }
+                    OrderEventLabel::Executing => {
+                        dto::order::Status::Executing(fetch_solutions().await?)
+                    }
+                    OrderEventLabel::Traded => dto::order::Status::Traded(fetch_solutions().await?),
```

Comment:
This is problematic as well, it depends how fast OnSettlementEventUpdater processed the settlement and races against new runloop iteration.

Reply:
Nice catch, thanks. Switched from the latest competition to the proposed query. Implemented using a single query to avoid multiple reads.